File: frontend/crm3/src/stores/product.ts
import { ref, computed } from 'vue'
import { defineStore } from 'pinia'

export interface Product {
  id?: number
  name: string
  description?: string
  sku: string
  price: number
  stock_quantity: number
  created_at?: string
  updated_at?: string
}

export interface ProductSearchParams {
  name?: string
  description?: string
  sku?: string
  min_price?: number
  max_price?: number
  page?: number
  limit?: number
}

export interface PaginatedProductResponse {
  products: Product[]
  total: number
  page: number
  limit: number
  total_pages: number
}

export const useProductStore = defineStore('product', () => {
  const products = ref<Product[]>([])
  const currentProduct = ref<Product | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)
  const pagination = ref({
    total: 0,
    page: 1,
    limit: 10,
    total_pages: 0
  })

  // Computed
  const isLoading = computed(() => loading.value)
  const hasError = computed(() => error.value !== null)
  const isEmpty = computed(() => products.value.length === 0)

  // Actions
  const clearError = () => {
    error.value = null
  }

  const setLoading = (value: boolean) => {
    loading.value = value
  }

  const setError = (message: string) => {
    error.value = message
    loading.value = false
  }

  const setProductsFromResponse = (data: PaginatedProductResponse) => {
    products.value = data.products
    pagination.value = {
      total: data.total,
      page: data.page,
      limit: data.limit,
      total_pages: data.total_pages
    }
  }

  const setCurrentProduct = (product: Product | null) => {
    currentProduct.value = product
  }

  const addProduct = (product: Product) => {
    products.value.unshift(product)
    pagination.value.total += 1
  }

  const updateProduct = (updatedProduct: Product) => {
    const index = products.value.findIndex(p => p.id === updatedProduct.id)
    if (index !== -1) {
      products.value[index] = updatedProduct
    }
    if (currentProduct.value?.id === updatedProduct.id) {
      currentProduct.value = updatedProduct
    }
  }

  const removeProduct = (productId: number) => {
    products.value = products.value.filter(p => p.id !== productId)
    pagination.value.total -= 1
    if (currentProduct.value?.id === productId) {
      currentProduct.value = null
    }
  }

  // API Methods (モックとして実装、後でAPI呼び出しに置き換え)
  const fetchProducts = async (params: ProductSearchParams = {}) => {
    try {
      setLoading(true)
      clearError()

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 500)) // モック遅延

      const mockData: PaginatedProductResponse = {
        products: [
          {
            id: 1,
            name: 'サンプル製品A',
            description: '高品質な製品A',
            sku: 'SKU-001',
            price: 1500,
            stock_quantity: 100,
            created_at: '2024-01-01'
          },
          {
            id: 2,
            name: 'サンプル製品B',
            description: '人気の製品B',
            sku: 'SKU-002',
            price: 2800,
            stock_quantity: 50,
            created_at: '2024-01-02'
          }
        ],
        total: 2,
        page: params.page || 1,
        limit: params.limit || 10,
        total_pages: 1
      }

      setProductsFromResponse(mockData)
    } catch (err) {
      setError(err instanceof Error ? err.message : '製品一覧の取得に失敗しました')
    } finally {
      setLoading(false)
    }
  }

  const fetchProduct = async (id: number) => {
    try {
      setLoading(true)
      clearError()

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 300))

      const mockProduct: Product = {
        id: id,
        name: 'サンプル製品A',
        description: '高品質な製品Aです。長期間の耐久性と優れた性能を提供します。',
        sku: 'SKU-001',
        price: 1500,
        stock_quantity: 100,
        created_at: '2024-01-01'
      }

      setCurrentProduct(mockProduct)
      return mockProduct
    } catch (err) {
      setError(err instanceof Error ? err.message : '製品情報の取得に失敗しました')
      throw err
    } finally {
      setLoading(false)
    }
  }

  const createProduct = async (productData: Omit<Product, 'id' | 'created_at' | 'updated_at'>) => {
    try {
      setLoading(true)
      clearError()

      // バリデーション
      if (!productData.name.trim()) {
        throw new Error('製品名は必須です')
      }
      if (!productData.sku.trim()) {
        throw new Error('SKUは必須です')
      }
      if (productData.price <= 0) {
        throw new Error('価格は正数である必要があります')
      }
      if (productData.stock_quantity < 0) {
        throw new Error('在庫数量は0以上である必要があります')
      }

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 500))

      const newProduct: Product = {
        ...productData,
        id: Date.now(), // モックID
        created_at: new Date().toISOString()
      }

      addProduct(newProduct)
      return newProduct
    } catch (err) {
      setError(err instanceof Error ? err.message : '製品の作成に失敗しました')
      throw err
    } finally {
      setLoading(false)
    }
  }

  const updateProductData = async (id: number, productData: Omit<Product, 'id' | 'created_at' | 'updated_at'>) => {
    try {
      setLoading(true)
      clearError()

      // バリデーション
      if (!productData.name.trim()) {
        throw new Error('製品名は必須です')
      }
      if (!productData.sku.trim()) {
        throw new Error('SKUは必須です')
      }
      if (productData.price <= 0) {
        throw new Error('価格は正数である必要があります')
      }
      if (productData.stock_quantity < 0) {
        throw new Error('在庫数量は0以上である必要があります')
      }

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 500))

      const updatedProduct: Product = {
        ...productData,
        id: id,
        updated_at: new Date().toISOString()
      }

      updateProduct(updatedProduct)
      return updatedProduct
    } catch (err) {
      setError(err instanceof Error ? err.message : '製品情報の更新に失敗しました')
      throw err
    } finally {
      setLoading(false)
    }
  }

  const deleteProduct = async (id: number) => {
    try {
      setLoading(true)
      clearError()

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 300))

      removeProduct(id)
    } catch (err) {
      setError(err instanceof Error ? err.message : '製品の削除に失敗しました')
      throw err
    } finally {
      setLoading(false)
    }
  }

  const adjustStock = async (id: number, quantity: number, operation: 'add' | 'subtract') => {
    try {
      setLoading(true)
      clearError()

      if (quantity <= 0) {
        throw new Error('数量は正数である必要があります')
      }

      // TODO: 実際のAPI呼び出しに置き換える
      await new Promise(resolve => setTimeout(resolve, 300))

      const product = products.value.find(p => p.id === id)
      if (product) {
        const newQuantity = operation === 'add'
          ? product.stock_quantity + quantity
          : Math.max(0, product.stock_quantity - quantity)

        const updatedProduct: Product = {
          ...product,
          stock_quantity: newQuantity,
          updated_at: new Date().toISOString()
        }

        updateProduct(updatedProduct)
      }

      if (currentProduct.value?.id === id) {
        await fetchProduct(id) // 詳細画面のデータも更新
      }
    } catch (err) {
      setError(err instanceof Error ? err.message : '在庫調整に失敗しました')
      throw err
    } finally {
      setLoading(false)
    }
  }

  // Helper methods for testing and direct state manipulation
  const setProducts = (productList: Product[], page: number, pageSize: number, total: number) => {
    products.value = productList
    pagination.value = {
      total,
      page,
      limit: pageSize,
      total_pages: Math.ceil(total / pageSize)
    }
  }

  return {
    // State
    products,
    currentProduct,
    loading,
    error,
    pagination,

    // Computed properties (exported as getters for easier access in tests)
    isLoading,
    hasError,
    isEmpty,
    totalCount: computed(() => pagination.value.total),
    currentPage: computed(() => pagination.value.page),
    pageSize: computed(() => pagination.value.limit),
    totalPages: computed(() => pagination.value.total_pages),

    // Actions
    clearError,
    setLoading,
    setError,
    setProducts,
    setCurrentProduct,
    addProduct,
    updateProduct,
    removeProduct,
    fetchProducts,
    fetchProduct,
    createProduct,
    updateProductData,
    deleteProduct,
    adjustStock
  }
})

File: backend/crm3_api/models/product.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Product 製品モデル
type Product struct {
	ID            uint      `json:"id" gorm:"primaryKey"`
	Name          string    `json:"name" gorm:"not null" validate:"required,min=1,max=100"`
	Description   *string   `json:"description,omitempty" validate:"omitempty,max=1000"`
	SKU           string    `json:"sku" gorm:"uniqueIndex;not null" validate:"required,min=1,max=50"`
	Price         float64   `json:"price" gorm:"not null" validate:"required,gt=0,lte=99999999"`
	StockQuantity int       `json:"stock_quantity" gorm:"not null" validate:"gte=0,lte=999999"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

// ProductCreateRequest 製品作成リクエスト
type ProductCreateRequest struct {
	Name          string  `json:"name" validate:"required,min=1,max=100"`
	Description   *string `json:"description,omitempty" validate:"omitempty,max=1000"`
	SKU           string  `json:"sku" validate:"required,min=1,max=50"`
	Price         float64 `json:"price" validate:"required,gt=0,lte=99999999"`
	StockQuantity int     `json:"stock_quantity" validate:"gte=0,lte=999999"`
}

// ProductUpdateRequest 製品更新リクエスト
type ProductUpdateRequest struct {
	Name          string  `json:"name" validate:"required,min=1,max=100"`
	Description   *string `json:"description,omitempty" validate:"omitempty,max=1000"`
	SKU           string  `json:"sku" validate:"required,min=1,max=50"`
	Price         float64 `json:"price" validate:"required,gt=0,lte=99999999"`
	StockQuantity int     `json:"stock_quantity" validate:"gte=0,lte=999999"`
}

// StockAdjustmentRequest 在庫調整リクエスト
type StockAdjustmentRequest struct {
	Quantity  int    `json:"quantity" validate:"required,gt=0"`
	Operation string `json:"operation" validate:"required,oneof=add subtract"`
	Reason    string `json:"reason,omitempty" validate:"omitempty,max=500"`
}

// ProductSearchParams 製品検索パラメータ
type ProductSearchParams struct {
	Name     string  `query:"name"`
	SKU      string  `query:"sku"`
	MinPrice float64 `query:"min_price"`
	MaxPrice float64 `query:"max_price"`
	Page     int     `query:"page"`
	Limit    int     `query:"limit"`
}

// PaginatedProductResponse ページネーション付き製品レスポンス
type PaginatedProductResponse struct {
	Products   []Product `json:"products"`
	Total      int64     `json:"total"`
	Page       int       `json:"page"`
	Limit      int       `json:"limit"`
	TotalPages int       `json:"total_pages"`
}

// ToProduct ProductCreateRequestをProductに変換
func (req *ProductCreateRequest) ToProduct() Product {
	return Product{
		Name:          req.Name,
		Description:   req.Description,
		SKU:           req.SKU,
		Price:         req.Price,
		StockQuantity: req.StockQuantity,
	}
}

// ToProduct ProductUpdateRequestをProductに変換
func (req *ProductUpdateRequest) ToProduct() Product {
	return Product{
		Name:          req.Name,
		Description:   req.Description,
		SKU:           req.SKU,
		Price:         req.Price,
		StockQuantity: req.StockQuantity,
	}
}

// BeforeCreate 作成前のフック
func (p *Product) BeforeCreate(tx *gorm.DB) error {
	// SKUの重複チェック
	var count int64
	if err := tx.Model(&Product{}).Where("sku = ?", p.SKU).Count(&count).Error; err != nil {
		return err
	}
	if count > 0 {
		return gorm.ErrDuplicatedKey
	}
	return nil
}

// BeforeUpdate 更新前のフック
func (p *Product) BeforeUpdate(tx *gorm.DB) error {
	// SKUの重複チェック（自分以外）
	var count int64
	if err := tx.Model(&Product{}).Where("sku = ? AND id != ?", p.SKU, p.ID).Count(&count).Error; err != nil {
		return err
	}
	if count > 0 {
		return gorm.ErrDuplicatedKey
	}
	return nil
}

// AdjustStock 在庫数量を調整
func (p *Product) AdjustStock(quantity int, operation string) error {
	switch operation {
	case "add":
		newQuantity := p.StockQuantity + quantity
		if newQuantity > 999999 {
			return gorm.ErrInvalidValue
		}
		p.StockQuantity = newQuantity
	case "subtract":
		newQuantity := p.StockQuantity - quantity
		if newQuantity < 0 {
			p.StockQuantity = 0
		} else {
			p.StockQuantity = newQuantity
		}
	default:
		return gorm.ErrInvalidValue
	}
	return nil
}

// GetStockStatus 在庫ステータスを取得
func (p *Product) GetStockStatus() string {
	if p.StockQuantity == 0 {
		return "在庫切れ"
	} else if p.StockQuantity <= 10 {
		return "在庫少"
	}
	return "在庫十分"
}

// GetStockValue 在庫評価額を計算
func (p *Product) GetStockValue() float64 {
	return p.Price * float64(p.StockQuantity)
}